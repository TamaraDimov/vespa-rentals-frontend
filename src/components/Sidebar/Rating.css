.rating-heading {
  font-size: 20px;
  color: rgb(255, 64, 0);
}

.default-heading {
  font-size: 18px;
  color: gray;
}

.star {
  cursor: pointer;
  border: none;
  background: none;
  font-size: 24px;
}

.star.selected {
  font-size: 24px;
  color: #ffa500;
}

@keyframes celebration {
  0% {
    transform: rotate(0);
    opacity: 1;
  }

  50% {
    transform: rotate(360deg);
    opacity: 0.5;
  }

  100% {
    transform: rotate(0);
    opacity: 1;
  }
}

.celebration-animation {
  animation-name: celebration;
  animation-duration: 1s;
  animation-iteration-count: 3;
  animation-timing-function: linear;
}
