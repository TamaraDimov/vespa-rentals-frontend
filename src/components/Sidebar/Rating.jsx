import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';
import './Rating.css';

const StarRating = () => {
  const user = useSelector((state) => state.user.user);
  const [rating, setRating] = useState(() => {
    const storedRating = localStorage.getItem(
      `starRating_${(user && user.user?.data?.id) || ''}`,
    );
    return storedRating ? parseInt(storedRating, 10) : 0;
  });

  useEffect(() => {
    if (user && user.user && user.user.data && user.user.data.id) {
      localStorage.setItem(
        `starRating_${user.user.data.id}`,
        rating.toString(),
      );
    }
  }, [rating, user]);

  const handleRatingChange = (selectedRating) => {
    setRating(selectedRating);
  };

  useEffect(() => {
    if (rating === 5) {
      const starElements = document.querySelectorAll('.star');
      starElements.forEach((starElement) => {
        starElement.classList.add('celebration-animation');
      });

      setTimeout(() => {
        starElements.forEach((starElement) => {
          starElement.classList.remove('celebration-animation');
        });
      }, 3000);
    }
  }, [rating]);

  const Star = ({ selected, onClick }) => (
    <button
      type="button"
      className={`star ${selected ? 'selected' : ''}`}
      onClick={onClick}
    >
      {selected ? '★' : '☆'}
    </button>
  );

  Star.propTypes = {
    selected: PropTypes.bool.isRequired,
    onClick: PropTypes.func.isRequired,
  };

  return (
    <div
      style={{ display: 'flex', flexDirection: 'column', alignItems: 'center' }}
    >
      {rating ? (
        <h2 className="rating-heading">{`Rating: ${rating}`}</h2>
      ) : (
        <h2 className="default-heading">⭐ Rate App ⬇️</h2>
      )}
      <div style={{ display: 'flex', justifyContent: 'center' }}>
        {[1, 2, 3, 4, 5].map((value) => (
          <Star
            key={value}
            selected={value <= rating}
            onClick={() => handleRatingChange(value)}
          />
        ))}
      </div>
    </div>
  );
};

export default StarRating;
