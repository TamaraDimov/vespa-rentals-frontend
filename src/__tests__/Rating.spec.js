import '@testing-library/jest-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import Rating from '../components/Sidebar/Rating';
import 'mock-local-storage';

const mockStore = configureMockStore();
const testuser = {
  user: {
    data: {
      id: 1,
      email: 'kiflekifle23@gmail.com',
      created_at: '2023-09-22T13:51:50.089Z',
    },
    token:
      'eyJhbGciOiJIUzI1NiJ9.eyJqdGkiOiJlYjI2ZTA2Yy1lMTM1LTRmMmEtYTYwZi1iYmMwM2I1MGE2MmQiLCJzdWIiOiIxIiwic2NwIjoidXNlciIsImF1ZCI6bnVsbCwiaWF0IjoxNjk2MjI2NTUwLCJleHAiOjE3MDE0MTA1NTB9.g0XLlQ7sc-1r0KIevNXVKwwH61FzdXM_sB-tpS6WWK4',
  },
};
const store = mockStore({
  user: {
    isLoading: false,
    user: testuser,
    token: testuser.token,
  },
});
describe('Rating', () => {
  let mockLocalStorage;
  let originalLocalStorage;

  beforeAll(() => {
    originalLocalStorage = global.localStorage;

    mockLocalStorage = {
      getItem: jest.fn(),
      setItem: jest.fn(),
      removeItem: jest.fn(),
      clear: jest.fn(),
    };

    const localStorageMock = {
      ...mockLocalStorage,
      getItem: jest.fn().mockImplementation((key) => mockLocalStorage[key]),
      setItem: jest.fn().mockImplementation((key, value) => {
        mockLocalStorage[key] = value;
      }),
      removeItem: jest.fn().mockImplementation((key) => {
        delete mockLocalStorage[key];
      }),
      clear: jest.fn().mockImplementation(() => {
        mockLocalStorage = {};
      }),
    };

    Object.defineProperty(global, 'localStorage', {
      value: localStorageMock,
    });

    localStorageMock.getItem.mockImplementation(() => JSON.stringify(testuser));
  });

  afterAll(() => {
    Object.defineProperty(global, 'localStorage', {
      value: originalLocalStorage,
    });
  });

  let rating;
  beforeAll(async () => {
    rating = renderer
      .create(
        <Provider store={store}>
          <Router>
            <Rating />
          </Router>
        </Provider>,
      )
      .toJSON();
  });

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('should render correctly', () => {
    expect(rating).toBeTruthy();
  });

  it('should match the snapshot', () => {
    expect(rating).toMatchSnapshot();
  });
});
