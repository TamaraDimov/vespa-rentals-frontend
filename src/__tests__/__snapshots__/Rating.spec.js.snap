// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating should match the snapshot 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
    }
  }
>
  <h2
    className="default-heading"
  >
    ⭐ Rate App ⬇️
  </h2>
  <div
    style={
      Object {
        "display": "flex",
        "justifyContent": "center",
      }
    }
  >
    <button
      className="star "
      onClick={[Function]}
      type="button"
    >
      ☆
    </button>
    <button
      className="star "
      onClick={[Function]}
      type="button"
    >
      ☆
    </button>
    <button
      className="star "
      onClick={[Function]}
      type="button"
    >
      ☆
    </button>
    <button
      className="star "
      onClick={[Function]}
      type="button"
    >
      ☆
    </button>
    <button
      className="star "
      onClick={[Function]}
      type="button"
    >
      ☆
    </button>
  </div>
</div>
`;
